### PR TITLE
add assembly implementation of md5 for aarch64

### PR DESCRIPTION
**Pull request by AWSjswinney**
A similar change is being considered for the [openssl](https://github.com/openssl/openssl/pull/16928) project. Applying the same change here will ensure that customers using the AWS SDKs on Graviton will get the best S3 upload performance for objects requiring md5 hashing. For objects 8MB and larger the throughput improvement is 27% on Graviton.

### Description of changes: 
This change adds an assembly implementation of md5 for aarch64 to achieve better performance for md5 hashing.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by AWSjswinney** [comment]
Here is the before and after performance profile:
```
./bssl speed -filter MD5
Did 1078000 MD5 (16 bytes) operations in 1000611us (1077341.7 ops/sec): 17.2 MB/s
Did 270000 MD5 (256 bytes) operations in 1000766us (269793.3 ops/sec): 69.1 MB/s
Did 65000 MD5 (1350 bytes) operations in 1005463us (64646.8 ops/sec): 87.3 MB/s
Did 12000 MD5 (8192 bytes) operations in 1082089us (11089.7 ops/sec): 90.8 MB/s
Did 6138 MD5 (16384 bytes) operations in 1098087us (5589.7 ops/sec): 91.6 MB/

./bssl speed -filter MD5
Did 3121000 MD5 (16 bytes) operations in 1000007us (3120978.2 ops/sec): 49.9 MB/s
Did 1232000 MD5 (256 bytes) operations in 1000033us (1231959.3 ops/sec): 315.4 MB/s
Did 344000 MD5 (1350 bytes) operations in 1001456us (343499.9 ops/sec): 463.7 MB/s
Did 63000 MD5 (8192 bytes) operations in 1013862us (62138.6 ops/sec): 509.0 MB/s
Did 32000 MD5 (16384 bytes) operations in 1018824us (31408.8 ops/sec): 514.6 MB/s
```

**Comment by AWSjswinney** [comment]
I mistakenly tested the before speed with a debug build. Here is the speed test with a release build without the change:
```
Did 5301000 MD5 (16 bytes) operations in 1000033us (5300825.1 ops/sec): 84.8 MB/s
Did 1242000 MD5 (256 bytes) operations in 1000106us (1241868.4 ops/sec): 317.9 MB/s
Did 289000 MD5 (1350 bytes) operations in 1001540us (288555.6 ops/sec): 389.6 MB/s
Did 50000 MD5 (8192 bytes) operations in 1010692us (49471.1 ops/sec): 405.3 MB/s
Did 25000 MD5 (16384 bytes) operations in 1006058us (24849.5 ops/sec): 407.1 MB/s
```
And after the change:
```
Did 3121000 MD5 (16 bytes) operations in 1000007us (3120978.2 ops/sec): 49.9 MB/s
Did 1232000 MD5 (256 bytes) operations in 1000033us (1231959.3 ops/sec): 315.4 MB/s
Did 344000 MD5 (1350 bytes) operations in 1001456us (343499.9 ops/sec): 463.7 MB/s
Did 63000 MD5 (8192 bytes) operations in 1013862us (62138.6 ops/sec): 509.0 MB/s
Did 32000 MD5 (16384 bytes) operations in 1018824us (31408.8 ops/sec): 514.6 MB/s
```

This is a 26.4% speed increase for 16K blocks. That is right in line with the improvement measured on the openssl project.

**Comment by AWSjswinney** [comment]
I added some changes from the openssl PR. We would like to go ahead and push this change since openssl is only waiting for a final "rubber stamp" review. Let me know if you need me to rebase or squash the commits.

File: crypto/fipsmodule/md5/asm/md5-armv8.pl [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+
+use strict;
+
+my $code;
+
+#no warnings qw(uninitialized);
+# $output is the last argument if it looks like a file (it has an extension)
+# $flavour is the first argument if it doesn't look like a file
+my $output = $#ARGV >= 0 && $ARGV[$#ARGV] =~ m|\.\w+$| ? pop : undef;
+my $flavour = $#ARGV >= 0 && $ARGV[0] !~ m|\.| ? shift : undef;
```
Comment (dkostic) [inline crypto/fipsmodule/md5/asm/md5-armv8.pl:19]:
```suggestion
# The first two arguments should always be the flavour and output file path.
if ($#ARGV < 1) { die "Not enough arguments provided.
  Two arguments are necessary: the flavour and the output file path."; }

my $flavour = shift;
my $output  = shift;
```

**Comment by dkostic** [comment]
> I mistakenly tested the before speed with a debug build. Here is the speed test with a release build without the change:
> 
> ```
> Did 5301000 MD5 (16 bytes) operations in 1000033us (5300825.1 ops/sec): 84.8 MB/s
> Did 1242000 MD5 (256 bytes) operations in 1000106us (1241868.4 ops/sec): 317.9 MB/s
> Did 289000 MD5 (1350 bytes) operations in 1001540us (288555.6 ops/sec): 389.6 MB/s
> Did 50000 MD5 (8192 bytes) operations in 1010692us (49471.1 ops/sec): 405.3 MB/s
> Did 25000 MD5 (16384 bytes) operations in 1006058us (24849.5 ops/sec): 407.1 MB/s
> ```
> 
> And after the change:
> 
> ```
> Did 6388000 MD5 (16 bytes) operations in 1000064us (6387591.2 ops/sec): 102.2 MB/s
> Did 1552000 MD5 (256 bytes) operations in 1000128us (1551801.4 ops/sec): 397.3 MB/s
> Did 366000 MD5 (1350 bytes) operations in 1002202us (365195.8 ops/sec): 493.0 MB/s
> Did 63000 MD5 (8192 bytes) operations in 1003583us (62775.1 ops/sec): 514.3 MB/s
> Did 32000 MD5 (16384 bytes) operations in 1013507us (31573.5 ops/sec): 517.3 MB/s
> ```
> 
> This is a 27.0% speed increase for 16K blocks. That is right in line with the improvement measured on the openssl project.

Is this measured on Graviton 2?

File: crypto/fipsmodule/md5/asm/md5-armv8.pl [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+*STDOUT=*OUT;
+
+$code .= <<EOF;
+
+.text
+.globl  md5_block_asm_data_order
+.type   md5_block_asm_data_order,\@function
+md5_block_asm_data_order:
+        // Save IP0, IP1, and all callee-saved registers
+        stp     x16,x17,[sp,#-96]!
```
Comment (dkostic) [inline crypto/fipsmodule/md5/asm/md5-armv8.pl:37]: as in the OpenSSL PR, x16 and x17 don't have to be saved.

**Comment by AWSjswinney** [comment]
> > I mistakenly tested the before speed with a debug build. Here is the speed test with a release build without the change:
> > ```
> > Did 5301000 MD5 (16 bytes) operations in 1000033us (5300825.1 ops/sec): 84.8 MB/s
> > Did 1242000 MD5 (256 bytes) operations in 1000106us (1241868.4 ops/sec): 317.9 MB/s
> > Did 289000 MD5 (1350 bytes) operations in 1001540us (288555.6 ops/sec): 389.6 MB/s
> > Did 50000 MD5 (8192 bytes) operations in 1010692us (49471.1 ops/sec): 405.3 MB/s
> > Did 25000 MD5 (16384 bytes) operations in 1006058us (24849.5 ops/sec): 407.1 MB/s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And after the change:
> > ```
> > Did 6388000 MD5 (16 bytes) operations in 1000064us (6387591.2 ops/sec): 102.2 MB/s
> > Did 1552000 MD5 (256 bytes) operations in 1000128us (1551801.4 ops/sec): 397.3 MB/s
> > Did 366000 MD5 (1350 bytes) operations in 1002202us (365195.8 ops/sec): 493.0 MB/s
> > Did 63000 MD5 (8192 bytes) operations in 1003583us (62775.1 ops/sec): 514.3 MB/s
> > Did 32000 MD5 (16384 bytes) operations in 1013507us (31573.5 ops/sec): 517.3 MB/s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is a 27.0% speed increase for 16K blocks. That is right in line with the improvement measured on the openssl project.
> 
> Is this measured on Graviton 2?

Yes. This is measured on Graviton 2.

**Comment by dkostic** [comment]
hey @AWSjswinney, I checked the CI failure logs. It looks like you just need to rerun the `util/generate_build_files.py` script because you modified the Perl assembly file in a previous commit.

**Comment by AWSjswinney** [comment]
> hey @AWSjswinney, I checked the CI failure logs. It looks like you just need to rerun the `util/generate_build_files.py` script because you modified the Perl assembly file in a previous commit.

Thanks... I always forget that. I have updated the generated source, but I still see some failures. What do you need from me next?

**Comment by torben-hansen** [comment]
Need to rebase or merge latest changes from main branch into your feature branch.

**Review by dkostic — APPROVED** [review]

**Review by nebeid — APPROVED** [review]

**Pull request merged**